lib/membrane_webm/muxer.ex

```diff
@@ -0,0 +1,260 @@
+    @moduledoc false
+    defstruct tracks: %{},
+              segment_size: 0,
+              expected_tracks: 0,
+              active_tracks: 0,
+              accepting_pads: true,
+              # cluster_acc holds: `serialized_list_of_blocks`, `cluster_timestamp`, `cluster_length_in_bytes`
+              cluster_acc: {Helper.serialize({:Timecode, 0}), :infinity, 0},
+              cues: []
+  end
+
+  @impl true
+  def handle_init(_options) do
+    {:ok, %State{}}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, state) do
+    {:ok, %State{state | accepting_pads: false}}
+  end
+
+  @impl true
+  def handle_pad_added(_pad, _context, state) do
+    if state.accepting_pads do
+      {:ok, %State{state | expected_tracks: state.expected_tracks + 1}}
+    else
+      raise "Can only add new input pads to muxer in state :prepared!"
+    end
+  end
+
+  @impl true
+  def handle_caps(Pad.ref(:input, id), caps, _context, state) do
+    codec =
+      case caps do
+        %VP8{} -> :vp8
+        %VP9{} -> :vp9
+        %Opus{} -> :opus
+        _other -> raise "unsupported codec #{inspect(caps)}"
+      end
+
+    state = update_in(state.active_tracks, fn t -> t + 1 end)
+    type = Codecs.type(codec)
+
+    track = %{
+      track_number: state.active_tracks,
+      id: id,
+      codec: codec,
+      caps: caps,
+      type: type,
+      last_timestamp: nil,
+      last_block: nil
+    }
+
+    state = put_in(state.tracks[id], track)
+
+    if state.active_tracks == state.expected_tracks do
+      {segment_size, webm_header} = Serializer.WebM.serialize_webm_header(state.tracks)
+      new_state = %State{state | segment_size: state.segment_size + segment_size}
+
+      {{:ok, [buffer: {:output, %Buffer{payload: webm_header}}, demand: Pad.ref(:input, id)]},
+       new_state}
+    else
+      {{:ok, demand: Pad.ref(:input, id)}, state}
+    end
+  end
+
+  # demand all tracks for which the last frame is not cached
+  @impl true
+  def handle_demand(:output, _size, :buffers, _context, state) do
+    if state.expected_tracks == state.active_tracks do
+      demands =
+        state.tracks
+        |> Enum.filter(fn {_id, info} -> info.last_block == nil end)
+        |> Enum.map(fn {id, _info} -> {:demand, Pad.ref(:input, id)} end)
+
+      {{:ok, demands}, state}
+    else
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_process(Pad.ref(:input, id), %Buffer{payload: data} = buffer, _context, state) do
+    timestamp = Buffer.get_dts_or_pts(buffer) |> div(@timestamp_scale)
+    state = put_in(state.tracks[id].last_timestamp, timestamp)
+    block = {timestamp, data, state.tracks[id].track_number, state.tracks[id].codec}
+    state = put_in(state.tracks[id].last_block, block)
+
+    if Enum.any?(state.tracks, fn {_id, info} -> info.last_block == nil end) do
+      # if there are active tracks without a cached frame then demand it
+      {{:ok, redemand: :output}, state}
+    else
+      consume_block(state)
```

Comment:
This is really confusing. You are consuming the block but then this element is a muxer... Are you sure that you're not creating a block?